File: app/src/main/java/com/simplemobiletools/draw/Svg.java
package com.simplemobiletools.draw;

import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.sax.Element;
import android.sax.RootElement;
import android.sax.StartElementListener;
import android.util.Xml;

import com.simplemobiletools.draw.actions.Action;

import org.xml.sax.Attributes;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStreamWriter;
import java.io.Serializable;
import java.io.Writer;
import java.util.ArrayList;
import java.util.Map;

public class Svg {

    //region Saving

    public static void saveSvg(File output, MyCanvas canvas) throws Exception {
        int backgroundColor = ((ColorDrawable) canvas.getBackground()).getColor();

        FileOutputStream out = new FileOutputStream(output);
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(out));
        writeSvg(writer, backgroundColor, canvas.getPaths(), canvas.getWidth(), canvas.getHeight());
        writer.close();
    }

    private static void writeSvg(Writer writer, int backgroundColor, Map<MyPath, PaintOptions> paths, int width, int height) throws IOException {
        writer.write("<svg width=\"");
        writer.write(String.valueOf(width));
        writer.write("\" height=\"");
        writer.write(String.valueOf(height));
        writer.write("\" xmlns=\"http://www.w3.org/2000/svg\">");

        // background rect
        writer.write("<rect width=\"");
        writer.write(String.valueOf(width));
        writer.write("\" height=\"");
        writer.write(String.valueOf(height));
        writer.write("\" fill=\"#");
        writer.write(Integer.toHexString(backgroundColor).substring(2)); // Skip the alpha FF
        writer.write("\"/>");

        for (Map.Entry<MyPath, PaintOptions> entry : paths.entrySet()) {
            writePath(writer, entry.getKey(), entry.getValue());
        }
        writer.write("</svg>");
    }

    private static void writePath(Writer writer, MyPath path, PaintOptions options) throws IOException {
        writer.write("<path d=\"");
        for (Action action : path.getActions()) {
            action.perform(writer);
            writer.write(' ');
        }

        writer.write("\" fill=\"none\" stroke=\"#");
        writer.write(Integer.toHexString(options.color).substring(2)); // Skip the alpha FF
        writer.write("\" stroke-width=\"");
        writer.write(String.valueOf(options.strokeWidth));
        writer.write("\" stroke-linecap=\"round\"/>");
    }

    //endregion

    //region Loading

    public static void loadSvg(File file, MyCanvas canvas) throws Exception {
        SSvg svg = parseSvg(file);

        canvas.clearCanvas();
        canvas.setBackgroundColor(svg.background.color);

        for (SPath sp : svg.paths) {
            MyPath path = new MyPath();
            path.readObject(sp.data);
            PaintOptions options = new PaintOptions(sp.color, sp.strokeWidth);

            canvas.addPath(path, options);
        }
    }

    public static SSvg parseSvg(File file) throws Exception {
        InputStream is = null;
        final SSvg svg = new SSvg();
        try {
            is = new FileInputStream(file);

            // Actual parsing (http://stackoverflow.com/a/4828765)
            final String ns = "http://www.w3.org/2000/svg";
            RootElement root = new RootElement(ns, "svg");
            Element rectElement = root.getChild(ns, "rect");
            final Element pathElement = root.getChild(ns, "path");

            root.setStartElementListener(new StartElementListener() {
                @Override
                public void start(Attributes attributes) {
                    int width = Integer.parseInt(attributes.getValue("width"));
                    int height = Integer.parseInt(attributes.getValue("height"));
                    svg.setSize(width, height);
                }
            });

            rectElement.setStartElementListener(new StartElementListener() {
                @Override
                public void start(Attributes attributes) {
                    int width = Integer.parseInt(attributes.getValue("width"));
                    int height = Integer.parseInt(attributes.getValue("height"));
                    int color = Color.parseColor(attributes.getValue("fill"));
                    if (svg.background != null)
                        throw new UnsupportedOperationException("Unsupported SVG, should only have one <rect>.");

                    svg.background = new SRect(width, height, color);
                }
            });

            pathElement.setStartElementListener(new StartElementListener() {
                public void start(Attributes attributes) {
                    String d = attributes.getValue("d");
                    int color = Color.parseColor(attributes.getValue("stroke"));
                    float width = Float.parseFloat(attributes.getValue("stroke-width"));
                    svg.paths.add(new SPath(d, color, width));
                }
            });

            // Once the parsing is set up, parse this InputStream
            Xml.parse(is, Xml.Encoding.UTF_8, root.getContentHandler());
        } finally {
            if (is != null)
                is.close();
        }
        return svg;
    }

    //region Svg serializable classes

    private static class SSvg implements Serializable {
        int width;
        int height;
        SRect background;
        final ArrayList<SPath> paths;

        SSvg() {
            paths = new ArrayList<>();
        }

        void setSize(int w, int h) {
            width = w;
            height = h;
        }
    }

    private static class SRect implements Serializable {
        final int width;
        final int height;
        final int color;

        SRect(int w, int h, int c) {
            width = w;
            height = h;
            color = c;
        }
    }

    private static class SPath implements Serializable {
        String data;
        int color;
        float strokeWidth;

        SPath(String d, int c, float w) {
            data = d;
            color = c;
            strokeWidth = w;
        }
    }

    //endregion

    //endregion
}

File: app/src/main/kotlin/com/simplemobiletools/draw/Svg.kt
package com.simplemobiletools.draw

import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.sax.RootElement
import android.util.Xml
import java.io.*
import java.util.*

object Svg {
    fun saveSvg(output: File, canvas: MyCanvas) {
        val backgroundColor = (canvas.background as ColorDrawable).color

        val out = FileOutputStream(output)
        val writer = BufferedWriter(OutputStreamWriter(out))
        writeSvg(writer, backgroundColor, canvas.mPaths, canvas.width, canvas.height)
        writer.close()
    }

    private fun writeSvg(writer: Writer, backgroundColor: Int, paths: Map<MyPath, PaintOptions>, width: Int, height: Int) {
        writer.apply {
            write("<svg width=\"")
            write(width.toString())
            write("\" height=\"")
            write(height.toString())
            write("\" xmlns=\"http://www.w3.org/2000/svg\">")

            // background rect
            write("<rect width=\"")
            write(width.toString())
            write("\" height=\"")
            write(height.toString())
            write("\" fill=\"#")
            write(Integer.toHexString(backgroundColor).substring(2))
            write("\"/>")

            for ((key, value) in paths) {
                writePath(this, key, value)
            }
            write("</svg>")
        }
    }

    private fun writePath(writer: Writer, path: MyPath, options: PaintOptions) {
        writer.apply {
            write("<path d=\"")
            path.getActions().forEach {
                it.perform(this)
                write(" ")
            }

            write("\" fill=\"none\" stroke=\"#")
            write(Integer.toHexString(options.color).substring(2))
            write("\" stroke-width=\"")
            write(options.strokeWidth.toString())
            write("\" stroke-linecap=\"round\"/>")
        }
    }

    fun loadSvg(file: File, canvas: MyCanvas) {
        val svg = parseSvg(file)

        canvas.clearCanvas()
        canvas.setBackgroundColor(svg.background!!.color)

        svg.paths.forEach {
            val path = MyPath()
            path.readObject(it.data)
            val options = PaintOptions(it.color, it.strokeWidth)

            canvas.addPath(path, options)
        }
    }

    private fun parseSvg(file: File): SSvg {
        var inputStream: InputStream? = null
        val svg = SSvg()
        try {
            inputStream = FileInputStream(file)

            // Actual parsing (http://stackoverflow.com/a/4828765)
            val ns = "http://www.w3.org/2000/svg"
            val root = RootElement(ns, "svg")
            val rectElement = root.getChild(ns, "rect")
            val pathElement = root.getChild(ns, "path")

            root.setStartElementListener { attributes ->
                val width = attributes.getValue("width").toInt()
                val height = attributes.getValue("height").toInt()
                svg.setSize(width, height)
            }

            rectElement.setStartElementListener { attributes ->
                val width = attributes.getValue("width").toInt()
                val height = attributes.getValue("height").toInt()
                val color = Color.parseColor(attributes.getValue("fill"))
                if (svg.background != null)
                    throw UnsupportedOperationException("Unsupported SVG, should only have one <rect>.")

                svg.background = SRect(width, height, color)
            }

            pathElement.setStartElementListener { attributes ->
                val d = attributes.getValue("d")
                val color = Color.parseColor(attributes.getValue("stroke"))
                val width = attributes.getValue("stroke-width").toFloat()
                svg.paths.add(SPath(d, color, width))
            }

            Xml.parse(inputStream, Xml.Encoding.UTF_8, root.contentHandler)
        } finally {
            inputStream?.close()
        }
        return svg
    }

    private class SSvg : Serializable {
        var background: SRect? = null
        val paths: ArrayList<SPath> = ArrayList()
        private var width = 0
        private var height = 0

        internal fun setSize(w: Int, h: Int) {
            width = w
            height = h
        }
    }

    private class SRect(val width: Int, val height: Int, val color: Int) : Serializable

    private class SPath(var data: String, var color: Int, var strokeWidth: Float) : Serializable
}
